Add support for new osbuild stages needed for GCE image

thozza opened the pull request:
- osbuild2: support `PermitRootLogin` in sshd.config stage
- osbuild2: support `org.osbuild.yum.repos` stage
- osbuild2: support `org.osbuild.dnf-automatic.config` stage

This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`
-->

msehnout commented on internal/osbuild2/yum_repos_stage.go:
```diff
@@ -0,0 +1,106 @@
+
+func (YumReposStageOptions) isStageOptions() {}
+
+// NewPamLimitsConfStageOptions creates a new PamLimitsConf Stage options object.
```
Copy paste error :-)

msehnout replied:
Suggestion: write a generator for these files. I think we are slowly approaching the point where the code is almost identical in the stage files. At least that was my impression after writing few of them in a row.

achilleas-k replied:
This comes up from time to time.  We could generate go code based on the json schema, which does weird things, but we wouldn't have to look at it :D 
A bit like the API autogeneration really.

thozza replied:
fixed the comment.

@achilleas-k I could probably live with that, as long as it works 😄

achilleas-k reviewed (dismissed):
Looks good.  Just some style comments and cleanup but approved.

msehnout commented on internal/osbuild2/dnf_automatic_config_stage.go:
```diff
@@ -0,0 +1,70 @@
+
+func (DNFAutomaticConfigStageOptions) isStageOptions() {}
+
+// NewPamLimitsConfStageOptions creates a new PamLimitsConf Stage options object.
```
same copy paste issue here

thozza replied:
removed

thozza commented on internal/osbuild2/sshd_config_stage.go:
```diff
@@ -1,9 +1,82 @@
+	var permitRootLogin PermitRootLoginValue
+	if rawConfig.PermitRootLogin != nil {
+		switch valueType := reflect.TypeOf(rawConfig.PermitRootLogin); valueType.Kind() {
+		// json.Unmarshal() uses float64 for JSON numbers
+		// https://pkg.go.dev/encoding/json#Unmarshal
+		// However the expected value is only integer.
```
oh yes... 🤦

thozza replied:
removed

msehnout reviewed (approved):
Looks pretty good! I have few non-critical suggestions.

thozza commented on internal/osbuild2/sshd_config_stage.go:
```diff
@@ -1,9 +1,82 @@
+
+	var permitRootLogin PermitRootLoginValue
+	if rawConfig.PermitRootLogin != nil {
+		switch valueType := reflect.TypeOf(rawConfig.PermitRootLogin); valueType.Kind() {
```
This should work and I fully agree with the readability. I'll fix this and also one other stage in which I used the same construct. Thanks for the suggestion.

thozza replied:
I made the change and also fixed the `pam.limits.conf` stage implementation in the same way...

thozza commented on internal/osbuild2/sshd_config_stage.go:
```diff
@@ -1,9 +1,82 @@
+// Valid values which can be used for the 'PermitRootLogin' item in
+// the SshdConfigConfig structure.
+const (
+	PermitRootLoginValueYes PermitRootLoginValueBool = true
```
A very valid question. This was in fact discussed in the related osbuild PR and the reason for using bool was to keep consistency with other options in the stage which do not accept string values and use only bool for "yes" / "no". More info is in https://github.com/osbuild/osbuild/pull/917

achilleas-k replied:
We have done this before where the stage in osbuild supports more formats or types than the one in composer.  I think in this case we decided it would be strange if booleans weren't supported at all.

thozza commented on internal/osbuild2/sshd_config_stage.go:
```diff
@@ -1,9 +1,82 @@
+	var permitRootLogin PermitRootLoginValue
+	if rawConfig.PermitRootLogin != nil {
+		switch valueType := reflect.TypeOf(rawConfig.PermitRootLogin); valueType.Kind() {
+		// json.Unmarshal() uses float64 for JSON numbers
```
removed

thozza commented:
I addressed all the comments and on top of that I also fixed the type switch statement in `pam.limits.conf` stage. Rebased on top of `main` and force-pushed...
